numpyro/infer/autoguide.py

```diff
@@ -50,13 +52,39 @@ class AutoGuide(ABC):
 
     :param callable model: a pyro model
     :param str prefix: a prefix that will be prefixed to all param internal sites
+    :param callable init_strategy: A per-site initialization function.
+        See :ref:`init_strategy` section for available functions.
+    :param callable create_plates: An optional function inputing the same
+        ``*args,**kwargs`` as ``model()`` and returning a :class:`numpyro.plate`
+        or iterable of plates. Plates not returned will be created
+        automatically as usual. This is useful for data subsampling.
     """
 
-    def __init__(self, model, prefix='auto'):
-        assert isinstance(prefix, str)
+    def __init__(self, model, *, prefix="auto", init_strategy=init_to_uniform, create_plates=None):
         self.model = model
         self.prefix = prefix
+        self.init_strategy = init_strategy
+        self.create_plates = create_plates
         self.prototype_trace = None
+        self._prototype_frames = {}
+        self._prototype_frame_full_sizes = {}
+
+    def _create_plates(self, *args, **kwargs):
+        if self.create_plates is None:
+            self.plates = {}
+        else:
+            plates = self.create_plates(*args, **kwargs)
+            if isinstance(plates, numpyro.plate):
+                plates = [plates]
+            assert all(isinstance(p, numpyro.plate) for p in plates), \
+                "create_plates() returned a non-plate"
+            self.plates = {p.name: p for p in plates}
+        for name, frame in sorted(self._prototype_frames.items()):
+            if name not in self.plates:
+                full_size = self._prototype_frame_full_sizes[name]
```

Comment:
In Pyro, we set `full_size` to be `frame.size` here. Is it a mistake? Without this change, I can't run SVI for the following model in #743
```python
        >>> def model(x, y=None, batch_size=None):
        ...     module = Net.partial(n_units=32)
        ...     net = random_flax_module("nn", module, dist.Normal(0, 0.1), input_shape=())
        ...     with numpyro.plate("batch", x.shape[0], subsample_size=batch_size):
        ...         batch_x = numpyro.subsample(x, event_dim=0)
        ...         batch_y = numpyro.subsample(y, event_dim=0) if y is not None else None
        ...         mean, rho = net(batch_x)
        ...         sigma = nn.softplus(rho)
        ...         numpyro.sample("obs", dist.Normal(mean, sigma), obs=batch_y)
```
unless I added `create_plates` to the guide.

Reply:
It looks like in Pyro we set `full_size = getattr(frame, "full_size", frame.size)` [here](https://github.com/pyro-ppl/pyro/blob/feeee7bcae6aefee6cca089f0d452e191aaf7535/pyro/infer/autoguide/guides.py#L347), and that the `full_size` attribute is set by `SubsampleMessenger` [here](https://github.com/pyro-ppl/pyro/blob/feeee7bcae6aefee6cca089f0d452e191aaf7535/pyro/poutine/subsample_messenger.py#L126). I think you'll also need the `frame.full_size` for param subsampling.

Reply:
Yeah, that Pyro code works for subsampled params. In the above model, we have subsampled data, so I think we need to use
```python
pyro.plate(name, full_size, dim=frame.dim, subsample_size=frame.size)
```
here instead of the current one in Pyro
```python
pyro.plate(name, frame.size, dim=frame.dim)
```
so in case `create_plates` is None, the `plate` in guide will be the same as the `plate` in model. (without this, I think `guide` will always generate `subsample=torch.arange(subsample_size)` and that constant index tensor will be used for subsampling in `model`)

Reply:
Thanks for explaining, I think I'm starting to understand. I believe historically Pyro only allowed guide subsampling via user-defined `create_plates()` functions, and ignored model subsampling (or mis-handled it as you suggest). Do I understand correctly that in your fixed version, the model's `subsample_size` will now determine the autoguide's strategy for automatic subsampling (rather than manual subsampling via `create_plates()`)?

I'll try to implement your fix in Pyro to help me understand. PR forthcoming 😄

Reply:
Here's your fix in Pyro 😄 https://github.com/pyro-ppl/pyro/pull/2638

Reply:
Yes, I believe you got it right. :D